draft/ZEP0002.md

```diff
@@ -0,0 +1,285 @@
+The index always has the full shape of all possible chunks per shard,
+even if they are outside of the array size.
+
+> The actual order of the chunk-content is not fixed and may be chosen by the implementation
+as all possible write orders are valid according to this specification and therefore can
+be read by any other implementation. When writing partial chunks into an existing shard no
+specific order of the existing chunks may be expected. Some writing strategies might be
+
+> * **Fixed order**: Specify a fixed order (e.g. row-, column-major or Morton order).
+    When replacing existing chunks larger or equal sized chunks may be replaced in-place,
+    leaving unused space up to an upper limit  which might possibly be specified.
+    Please note that for regular-sized uncompressed data all chunks have the same size and
+    can therefore be replaced in-place.
+> * **Append-only**: Any chunk to write is appended to the existing shard,
+    followed by an updated index.
+
+> Any configuration parameters for the write strategy must not be part of the metadata document,
+they need to be configured at runtime, as this is implementation specific.
+
+
+### Use-cases and examples
+
+
+#### Reading data per chunk
+
+Reading single chunks can be done in two ways, depending on the capabilities of the underlying storage:
+1. * Downloading the complete corresponding shard,
+   * reading the index and
+   * afterwards reading the relevant chunk.
+2. * Downloading only the index of the corresponding shard,
+   * identifying the relevant byte-range for the requested chunk,
+   * only downloading this byte-range from the same shard.
+
+The first method might especially be useful when caching not only the requested chunks but the whole shard,
+which will save bandwidth if subsequent chunk-requests are close to the previous chunks.
+
+#### Storing a stream of data slices
+
+Many image acquisition techniques produce slice-wise data, such as timeseries or z-slicing. This data can
+be written slice-wise using one of the following techniques:
+1. Using a chunk-size and shard-size of 1 for the sliced dimension effectively disables chunking and sharding for it.
+2. Slice-wise data with a chunk-size of 1 for the sliced dimension can be appended to a shard, only wasting the space
+   for the index that was written before.
+3. When using no compression, chunk-sizes are constant, and shard indices can be known ahead of time, allowing
+   to write partial chunks of a shard. Therefore, a chunk-size of 1 can be used for the sliced dimension, in addition
+   to a higher shard-size for this dimension. This is beneficial if later access typically constists of blocks with
+   multiple slices.
+4. Strategy 2 can be combined with compression by either
+   * compressing chunks after all slices of a "shard-layer" are written, or
+   * directly writing compressed chunks, but assuming a constant chunk-size for the index,
```

Comment:
No need to waste any space, when writing a new shard index after appending a chunk:
`chunk_data ... index` -> `chunk_data ... new_chunks ... new_index`

Reply:
Good point, somehow I was assuming an append-only store.